src/main/java/org/zalando/nakadi/service/subscription/KafkaClient.java

```diff
@@ -1,57 +1,88 @@
         this.subscription = subscription;
-        this.eventTypeRepository = eventTypeRepository;
+        this.timelineService = timelineService;
+        this.cursorConverter = cursorConverter;
         // FIXME TIMELINE: for refactoring purposes, has to be removed during timeline event consumption task
         // for now we always will have the same topic repo, KafkaTopicRepository
-        try {
-            final EventType eventType = eventTypeRepository.findByName(subscription.getEventTypes().iterator().next());
-            this.topicRepository = (KafkaTopicRepository) timelineService.getTopicRepository(eventType);
-        } catch (final NakadiException e) {
-            throw new NakadiRuntimeException(e);
-        }
+        this.topicRepository = timelineService.getTopicRepository(
+                timelineService.getActiveTimelinesOrdered(subscription.getEventTypes().iterator().next()).get(0));
         // FIXME TIMELINE: for refactoring purposes, has to be removed during timeline event consumption task
     }
 
-    public Map<Partition.PartitionKey, Long> getSubscriptionOffsets() {
-        final Map<Partition.PartitionKey, Long> offsets = new HashMap<>();
+    public Map<Partition.PartitionKey, NakadiCursor> getSubscriptionOffsets() {
         try {
-            for (final String eventTypeName : subscription.getEventTypes()) {
-                final EventType eventType = eventTypeRepository.findByName(eventTypeName);
-                final String topic = eventType.getTopic();
-                topicRepository.materializePositions(eventType, subscription)
-                        .entrySet()
-                        .forEach(
-                                e -> offsets.put(new Partition.PartitionKey(topic, e.getKey()), e.getValue() - 1));
+            final List<NakadiCursor> existingPositions = new ArrayList<>();
+            switch (subscription.getReadFrom()) {
+                case BEGIN:
+                    for (String eventType : subscription.getEventTypes()) {
+                        final Timeline timeline = timelineService.getActiveTimelinesOrdered(eventType).get(0);
+                        timelineService.getTopicRepository(timeline)
+                                .loadTopicStatistics(Collections.singletonList(timeline))
+                                .forEach(stat -> existingPositions.add(stat.getBeforeFirst()));
+                    }
+                    break;
+                case END:
+                    for (String eventType : subscription.getEventTypes()) {
+                        final Timeline timeline = timelineService.getActiveTimelinesOrdered(eventType).get(0);
+                        timelineService.getTopicRepository(timeline)
+                                .loadTopicStatistics(Collections.singletonList(timeline))
+                                .forEach(stat -> existingPositions.add(stat.getLast()));
+                    }
+                    break;
```

Comment:
branches for BEGIN and END are almost identical. I think the code can be improved to avoid duplication. For example, to check if it's BEGIN or END only when you do `stat.getBeforeFirst()` or `stat.getLast()`